In [1]:
%pip install pandas
%pip install -r requirements.txt

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [2]:
%pip install xgboost

Note: you may need to restart the kernel to use updated packages.


In [3]:
%pip install optuna

Note: you may need to restart the kernel to use updated packages.


In [4]:
import pandas as pd
import numpy as np
from sklearn.feature_selection import VarianceThreshold, SelectKBest, f_regression, f_classif, SelectFromModel
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, LassoCV, LogisticRegression
from sklearn.metrics import mean_squared_error, roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC

from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV

In [5]:
df = pd.read_csv("./datasets/train/train.csv") # для train

In [6]:

# Удаляем индекс, так как он не несёт информации для модели
X = df.drop(['index', 'target'], axis=1)  # Все фичи (feature_1 ... feature_1366)
y = df['target']  # Целевая переменная

X, y

(        feature_0  feature_1  feature_2  feature_3  feature_4  feature_5  \
 0        0.128367        1.0        1.0        1.0        1.0        4.0   
 1        0.258685        1.0        1.0        1.0        0.0        7.0   
 2        0.407738        1.0        1.0        1.0        2.0        7.0   
 3        0.439134        1.0        1.0        1.0        1.0        2.0   
 4        0.590481        1.0        1.0        1.0        0.0        1.0   
 ...           ...        ...        ...        ...        ...        ...   
 177853   0.660324        1.0        1.0        1.0        0.0        1.0   
 177854   0.222214        0.0        1.0        1.0        1.0        4.0   
 177855   0.545399        0.0        1.0        1.0        0.0        2.0   
 177856   0.393150        1.0        1.0        1.0        0.0        1.0   
 177857   0.395272        1.0        1.0        1.0        1.0        2.0   
 
         feature_6  feature_7  feature_8  feature_9  ...  feature_1357  \


In [7]:
indexes = df['index']

In [6]:
# 1. Отбор признаков: удаляем низковариативные (константные и почти константные)
# Сделаем pipeline для более эффективного отбора признаков(их сокращения)
pipe = Pipeline([
    ('selector', VarianceThreshold(threshold=0.01)),
    ('kbest', SelectKBest(f_classif, k=200)),
    # ('lasso_selection', SelectFromModel(LassoCV(cv=2), max_features=500))
])

X_selected = pipe.fit_transform(X, y)
selector = pipe.named_steps['kbest']

print(f"Исходное число признаков: {X.shape[1]}")
print(f"Число признаков после отбора: {X_selected.shape[1]}")

Исходное число признаков: 1367
,Число признаков после отбора: 200


In [7]:
# 2. Масштабирование данных (важно для линейной регрессии)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_selected)

X_train, X_test, y_train, y_test = train_test_split(
    X_selected, y, test_size=0.3, random_state=42
)


In [ ]:
# 4. Создание и обучение модели
model = LinearRegression()
model.fit(X_train, y_train)

# 5. Предсказание и оценка качества
y_pred = model.predict(X_test)
# y_pred_proba_lin_reg = model.decision_function(X_test)

# Метрики
mse = mean_squared_error(y_test, y_pred)
# нормализуем значения для roc_auc
y_proba = 1 / (1 + np.exp(-y_pred)) # преобразование через сигмойду
roc_auc = roc_auc_score(y_test, y_proba)

print(f"\nMSE: {mse:.4f}")
print(f"ROC-AUC Score: {roc_auc:.4f}")

# (Опционально) Вывод коэффициентов модели
if X_selected.shape[1] <= 20:  # Не выводить для слишком большого числа фич
    print("\nКоэффициенты модели:")
    # for feature, coef in zip(X.columns[selector.get_support()], model.coef_):
    #     print(f"{feature}: {coef:.4f}")
else:
    print(f"\nМодель имеет {X_selected.shape[1]} коэффициентов (слишком много для вывода).")


,MSE: 0.0171
,R2 Score: 0.0016
,ROC-AUC Score: 0.6536
,
,Модель имеет 400 коэффициентов (слишком много для вывода).


In [42]:
y_pred

array([ 0.01629677,  0.00887584,  0.00911322, ...,  0.0150569 ,
       -0.0037384 , -0.00022041], shape=(53358,))

Так.... Из полученных метрик следует что вроде бы по MSE результат довольно терпимый и среднеквадратичная ошибка не слишком большая.
Касаемо метрики ROC-AUC Score получаем 0.6536 что является лучше чем убычное угадывание(подбрасывание монетки).
Из инересного мы не должны использовать LinearRegression, так как она может выдавать значения за пределами [0, 1], но после регуляризации значений, они могут быть использованиы для вычисления метрики. 

Попробуем RandomForestClassifier

In [ ]:
model = RandomForestClassifier()
model.fit(X_train, y_train)

y_pred_proba = model.predict_proba(X_test)[:, 1]
roc_auc = roc_auc_score(y_test, y_pred_proba)


In [13]:
y_pred_proba

array([0.  , 0.04, 0.03, ..., 0.  , 0.03, 0.  ], shape=(53358,))

In [47]:
print(f"ROC_AUC_SCORE is {roc_auc:.4f}")

ROC_AUC_SCORE is 0.6261


Прогресс, модель достигла случайного угадывания
(последние данные ROC_AUC_SCORE is 0.6261, это даже лучше чем просто уадывание по среднему значению)

In [32]:
mse = mean_squared_error(y_test, y_pred_proba)
print(f'MSE: {mse:.2f}')

MSE: 0.02


Небольшой апдейт для структуры модели

In [12]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

In [ ]:
# 1. Отбор признаков: удаляем низковариативные (константные и почти константные)
# Сделаем pipeline для более эффективного отбора признаков(их сокращения)
pipe = Pipeline([
    ('selector', VarianceThreshold(threshold=0.01)),
    ('kbest', SelectKBest(f_classif, k=100)),
    ('scaler', StandardScaler()),
    ('model', RandomForestClassifier())
    # ('lasso_selection', SelectFromModel(LassoCV(cv=2), max_features=500))
])

pipe.fit(X_train, y_train)

y_pred = pipe.predict_proba(X_test)[:, 1]

roc_auc_metric = roc_auc_score(y_test, y_pred)


In [33]:
print(f"ROC-AUC Score is {roc_auc_metric:.4f}")

ROC-AUC Score is 0.6221


А что если заменить RandomForsetClassifier на LogisticRegression

In [ ]:
# 1. Отбор признаков: удаляем низковариативные (константные и почти константные)
# Сделаем pipeline для более эффективного отбора признаков(их сокращения)
pipe = Pipeline([
    ('selector', VarianceThreshold(threshold=0.01)),
    ('kbest', SelectKBest(f_classif, k=100)),
    ('scaler', StandardScaler()),
    ('model', LogisticRegression())
    # ('lasso_selection', SelectFromModel(LassoCV(cv=2), max_features=500))
])

pipe.fit(X_train, y_train)

y_pred = pipe.predict_proba(X_test)[:, 1]

roc_auc_metric = roc_auc_score(y_test, y_pred)

In [37]:
print(f"ROC-AUC Score is {roc_auc_metric:.4f}")

ROC-AUC Score is 0.6473


Вау, метрика подросла)


Теперь попробуем GradientBoostingClassifier


In [13]:
# 1. Отбор признаков: удаляем низковариативные (константные и почти константные)
# Сделаем pipeline для более эффективного отбора признаков(их сокращения)
pipe = Pipeline([
    ('selector', VarianceThreshold(threshold=0.01)),
    ('kbest', SelectKBest(f_classif, k=100)),
    ('scaler', StandardScaler()),
    ('model', GradientBoostingClassifier())
    # ('lasso_selection', SelectFromModel(LassoCV(cv=2), max_features=500))
])

pipe.fit(X_train, y_train)

y_pred = pipe.predict_proba(X_test)[:, 1]

roc_auc_metric = roc_auc_score(y_test, y_pred)

In [14]:
print(f"ROC-AUC Score is {roc_auc_metric:.4f}")

ROC-AUC Score is 0.6791


Попробуем еще одну модель SVC

In [6]:
# 1. Отбор признаков: удаляем низковариативные (константные и почти константные)
# Сделаем pipeline для более эффективного отбора признаков(их сокращения)
pipe = Pipeline([
    ('selector', VarianceThreshold(threshold=0.01)),
    ('kbest', SelectKBest(f_classif, k=100)),
    ('scaler', StandardScaler()),
    ('model', SVC(probability=True, kernel='rbf', C=1.0, decision_function_shape='ovr'))
    # ('lasso_selection', SelectFromModel(LassoCV(cv=2), max_features=500))
])

pipe.fit(X_train, y_train)

y_pred = pipe.predict_proba(X_test)[:, 1]

roc_auc_metric = roc_auc_score(y_test, y_pred)

In [7]:
print(f"ROC-AUC Score is {roc_auc_metric:.4f}")

ROC-AUC Score is 0.5873


Начинаем работу с XGBoost

Подготовим данные для него

In [8]:
pipe = Pipeline([
    ('selector', VarianceThreshold(threshold=0.01)),
    ('scaler', StandardScaler()),
    # ('kbest', SelectKBest(f_classif, k=500)),
])

X_selected = pipe.fit_transform(X, y)

In [9]:
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2, interaction_only=True)
X_poly = poly.fit_transform(X_selected)

MemoryError: Unable to allocate 1.06 TiB for an array with shape (177858, 821122) and data type float64

Для полиномиальных признаков

In [6]:
X_train, X_test, y_train, y_test = train_test_split(
    X_poly, y, test_size=0.2, random_state=42
)

NameError: name 'X_poly' is not defined

In [10]:
X_train, X_test, y_train, y_test = train_test_split(
    X_selected, y, test_size=0.2, random_state=42
)

Попытаемся создать полиномиальные признаки

Присутствуют импорты библиотек и моделей, чтобы лучше понимать что происходит и не возвращаться каждый раз наверх

Пипец долго считает)

In [ ]:
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV

model = XGBClassifier(objective='binary:logistic', eval_metric='auc', tree_method='hist', device='cuda')

param_grid = {
    'n_estimators': [500, 1000, 1500, 1700],
    'max_depth': [3, 6, 9],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.7, 0.9],
    'colsample_bytree': [0.7, 0.9],
    'reg_alpha': [0, 1],
    'reg_lambda': [0, 1]
}

grid = GridSearchCV(model, param_grid, scoring='roc_auc', cv=5, n_jobs=-1)
grid.fit(X_train, y_train)

print(f"Лучшие параметры: {grid.best_params_}")
print(f"Лучший ROC-AUC: {grid.best_score_:.4f}")

In [9]:
import optuna
from sklearn.metrics import roc_auc_score

def objective(trial):
    params = {
        'n_estimators': trial.suggest_int('n_estimators', 500, 3000),
        'max_depth': trial.suggest_int('max_depth', 3, 10),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.3),
        'subsample': trial.suggest_float('subsample', 0.7, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.7, 1.0),
        'reg_alpha': trial.suggest_float('reg_alpha', 0, 10),
        'reg_lambda': trial.suggest_float('reg_lambda', 0, 10),
    }
    
    model = XGBClassifier(**params, objective='binary:logistic', 
                          scale_pos_weight=10, device='cuda', tree_method='hist', eval_metric='auc')
    model.fit(X_train, y_train)
    y_proba = model.predict_proba(X_test)[:, 1]
    return roc_auc_score(y_test, y_proba)

study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50, show_progress_bar=True)
print(f"Лучший ROC-AUC: {study.best_value:.4f}")
print(f"Лучшие параметры: {study.best_params}")

[I 2025-04-24 15:47:06,315] A new study created in memory with name: no-name-ad40d7ca-dec2-4e7d-8ff9-560291bc2fde
,  0%|          | 0/50 [00:00<?, ?it/s]c:\Users\ADMIN\AppData\Local\Programs\Python\Python312\Lib\site-packages\xgboost\core.py:729: UserWarning: [15:48:44] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\common\error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.
,Potential solutions:
,- Use a data structure that matches the device ordinal in the booster.
,- Set the device for booster before call to inplace_predict.
,
,This warning will only be shown once.
,
,  return func(**kwargs)
,Best trial: 0. Best value: 0.719963:   2%|▏         | 1/50 [01:39<1:20:56, 99.11s/it]

[I 2025-04-24 15:48:45,425] Trial 0 finished with value: 0.7199628349060889 and parameters: {'n_estimators': 1125, 'max_depth': 8, 'learning_rate': 0.02241833424878355, 'subsample': 0.8096449737437666, 'colsample_bytree': 0.7073004384635065, 'reg_alpha': 6.533106038440083, 'reg_lambda': 9.262728769416047}. Best is trial 0 with value: 0.7199628349060889.


Best trial: 0. Best value: 0.719963:   4%|▍         | 2/50 [03:13<1:16:53, 96.12s/it]

[I 2025-04-24 15:50:19,449] Trial 1 finished with value: 0.6954017664553149 and parameters: {'n_estimators': 2776, 'max_depth': 4, 'learning_rate': 0.18976614883120557, 'subsample': 0.9024856139308948, 'colsample_bytree': 0.7919114539903954, 'reg_alpha': 0.8403179895039259, 'reg_lambda': 5.852740218876109}. Best is trial 0 with value: 0.7199628349060889.


Best trial: 0. Best value: 0.719963:   6%|▌         | 3/50 [05:59<1:40:17, 128.03s/it]

[I 2025-04-24 15:53:05,448] Trial 2 finished with value: 0.7180354475785997 and parameters: {'n_estimators': 1978, 'max_depth': 9, 'learning_rate': 0.05694786801733289, 'subsample': 0.7847523081993439, 'colsample_bytree': 0.7076873877312242, 'reg_alpha': 4.669880749825391, 'reg_lambda': 2.8157237485636477}. Best is trial 0 with value: 0.7199628349060889.


Best trial: 0. Best value: 0.719963:   8%|▊         | 4/50 [08:18<1:41:30, 132.41s/it]

[I 2025-04-24 15:55:24,573] Trial 3 finished with value: 0.713451473531361 and parameters: {'n_estimators': 2097, 'max_depth': 7, 'learning_rate': 0.09032277147875942, 'subsample': 0.9002062763042251, 'colsample_bytree': 0.837580096391722, 'reg_alpha': 3.7472901513237202, 'reg_lambda': 3.9775543268465596}. Best is trial 0 with value: 0.7199628349060889.


Best trial: 0. Best value: 0.719963:  10%|█         | 5/50 [09:31<1:23:15, 111.01s/it]

[I 2025-04-24 15:56:37,647] Trial 4 finished with value: 0.700522743565372 and parameters: {'n_estimators': 2614, 'max_depth': 3, 'learning_rate': 0.19727794585205108, 'subsample': 0.9200595290250553, 'colsample_bytree': 0.8780237782872266, 'reg_alpha': 6.324090389445636, 'reg_lambda': 3.1349286728369705}. Best is trial 0 with value: 0.7199628349060889.


Best trial: 0. Best value: 0.719963:  12%|█▏        | 6/50 [10:38<1:10:24, 96.02s/it] 

[I 2025-04-24 15:57:44,559] Trial 5 finished with value: 0.6937455807809569 and parameters: {'n_estimators': 2294, 'max_depth': 5, 'learning_rate': 0.1097006062085589, 'subsample': 0.976071773775874, 'colsample_bytree': 0.8239415761892432, 'reg_alpha': 0.6321924008529678, 'reg_lambda': 2.1126222794157923}. Best is trial 0 with value: 0.7199628349060889.


Best trial: 0. Best value: 0.719963:  14%|█▍        | 7/50 [11:58<1:05:06, 90.84s/it]

[I 2025-04-24 15:59:04,741] Trial 6 finished with value: 0.6935528610799031 and parameters: {'n_estimators': 2194, 'max_depth': 9, 'learning_rate': 0.2844124627876331, 'subsample': 0.9479362530810237, 'colsample_bytree': 0.7456803339686764, 'reg_alpha': 2.0283629974709774, 'reg_lambda': 9.166632427374378}. Best is trial 0 with value: 0.7199628349060889.


Best trial: 7. Best value: 0.723551:  16%|█▌        | 8/50 [12:58<56:39, 80.94s/it]  

[I 2025-04-24 16:00:04,465] Trial 7 finished with value: 0.7235508565666319 and parameters: {'n_estimators': 2390, 'max_depth': 5, 'learning_rate': 0.011511099779345708, 'subsample': 0.7274638837984257, 'colsample_bytree': 0.7870636515555891, 'reg_alpha': 2.597695779726802, 'reg_lambda': 5.34023274266019}. Best is trial 7 with value: 0.7235508565666319.


Best trial: 7. Best value: 0.723551:  18%|█▊        | 9/50 [13:28<44:25, 65.02s/it]

[I 2025-04-24 16:00:34,494] Trial 8 finished with value: 0.7164686157120631 and parameters: {'n_estimators': 1690, 'max_depth': 3, 'learning_rate': 0.08401808970530573, 'subsample': 0.9495905526162115, 'colsample_bytree': 0.7742410137355674, 'reg_alpha': 9.2542879876784, 'reg_lambda': 1.4110384779504381}. Best is trial 7 with value: 0.7235508565666319.


Best trial: 7. Best value: 0.723551:  20%|██        | 10/50 [14:28<42:17, 63.44s/it]

[I 2025-04-24 16:01:34,411] Trial 9 finished with value: 0.7075546693763993 and parameters: {'n_estimators': 1526, 'max_depth': 6, 'learning_rate': 0.12122429961263705, 'subsample': 0.9210130667271419, 'colsample_bytree': 0.9465049678910262, 'reg_alpha': 8.902489745746562, 'reg_lambda': 8.123063859232982}. Best is trial 7 with value: 0.7235508565666319.


Best trial: 7. Best value: 0.723551:  22%|██▏       | 11/50 [14:53<33:45, 51.94s/it]

[I 2025-04-24 16:02:00,250] Trial 10 finished with value: 0.721526740649526 and parameters: {'n_estimators': 746, 'max_depth': 6, 'learning_rate': 0.01243079431096733, 'subsample': 0.7173016874036109, 'colsample_bytree': 0.8948034278553091, 'reg_alpha': 2.879613365712022, 'reg_lambda': 6.077309585779324}. Best is trial 7 with value: 0.7235508565666319.


Best trial: 11. Best value: 0.726775:  24%|██▍       | 12/50 [15:16<27:19, 43.14s/it]

[I 2025-04-24 16:02:23,264] Trial 11 finished with value: 0.7267751825365547 and parameters: {'n_estimators': 653, 'max_depth': 6, 'learning_rate': 0.02176645480492359, 'subsample': 0.7021475108158397, 'colsample_bytree': 0.9160694536741326, 'reg_alpha': 2.5180991485275133, 'reg_lambda': 6.366469735868886}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  26%|██▌       | 13/50 [15:32<21:29, 34.86s/it]

[I 2025-04-24 16:02:39,090] Trial 12 finished with value: 0.7245863235099813 and parameters: {'n_estimators': 533, 'max_depth': 5, 'learning_rate': 0.04216421206782048, 'subsample': 0.7029768264100341, 'colsample_bytree': 0.9783963562377573, 'reg_alpha': 1.9301392767346173, 'reg_lambda': 6.8168685384749015}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  28%|██▊       | 14/50 [16:05<20:30, 34.17s/it]

[I 2025-04-24 16:03:11,650] Trial 13 finished with value: 0.7020038376461295 and parameters: {'n_estimators': 732, 'max_depth': 7, 'learning_rate': 0.05899633962464237, 'subsample': 0.7634272518482134, 'colsample_bytree': 0.9903073627273128, 'reg_alpha': 0.07941937955347167, 'reg_lambda': 7.202462721893033}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  30%|███       | 15/50 [16:21<16:42, 28.65s/it]

[I 2025-04-24 16:03:27,499] Trial 14 finished with value: 0.7005981804462562 and parameters: {'n_estimators': 552, 'max_depth': 5, 'learning_rate': 0.14572867805895473, 'subsample': 0.8428372518885123, 'colsample_bytree': 0.9396165341888109, 'reg_alpha': 1.8093986416637193, 'reg_lambda': 7.2088390918986684}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  32%|███▏      | 16/50 [18:04<28:58, 51.12s/it]

[I 2025-04-24 16:05:10,813] Trial 15 finished with value: 0.7206591094660748 and parameters: {'n_estimators': 1205, 'max_depth': 10, 'learning_rate': 0.0518858914422769, 'subsample': 0.7096449586979756, 'colsample_bytree': 0.9980424295495613, 'reg_alpha': 4.238096339577529, 'reg_lambda': 4.476714254586341}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  34%|███▍      | 17/50 [18:31<24:11, 43.98s/it]

[I 2025-04-24 16:05:38,184] Trial 16 finished with value: 0.6955846134646102 and parameters: {'n_estimators': 1131, 'max_depth': 4, 'learning_rate': 0.21864001733299962, 'subsample': 0.7513907204041831, 'colsample_bytree': 0.9383574678075621, 'reg_alpha': 5.992597605950136, 'reg_lambda': 0.15343057431670815}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  36%|███▌      | 18/50 [18:49<19:16, 36.14s/it]

[I 2025-04-24 16:05:56,081] Trial 17 finished with value: 0.7134886329157636 and parameters: {'n_estimators': 508, 'max_depth': 6, 'learning_rate': 0.04111954387693036, 'subsample': 0.8231030927439897, 'colsample_bytree': 0.9008914016263138, 'reg_alpha': 3.363879223463165, 'reg_lambda': 6.775643165594782}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  38%|███▊      | 19/50 [19:24<18:31, 35.86s/it]

[I 2025-04-24 16:06:31,282] Trial 18 finished with value: 0.6800444494751559 and parameters: {'n_estimators': 940, 'max_depth': 4, 'learning_rate': 0.2877945809701882, 'subsample': 0.7535813834116671, 'colsample_bytree': 0.9664542472319517, 'reg_alpha': 1.5931881268015216, 'reg_lambda': 8.408771125295253}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  40%|████      | 20/50 [20:52<25:38, 51.27s/it]

[I 2025-04-24 16:07:58,469] Trial 19 finished with value: 0.7002075073280352 and parameters: {'n_estimators': 1416, 'max_depth': 8, 'learning_rate': 0.15155922523859075, 'subsample': 0.7960803023563698, 'colsample_bytree': 0.9161609668562454, 'reg_alpha': 7.467945677659381, 'reg_lambda': 4.938805363521544}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  42%|████▏     | 21/50 [21:51<25:54, 53.59s/it]

[I 2025-04-24 16:08:57,457] Trial 20 finished with value: 0.713798992281268 and parameters: {'n_estimators': 741, 'max_depth': 7, 'learning_rate': 0.07727741130779367, 'subsample': 0.7009051046610961, 'colsample_bytree': 0.8648459306012585, 'reg_alpha': 5.367475457372855, 'reg_lambda': 8.04953953326703}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  44%|████▍     | 22/50 [23:34<32:02, 68.67s/it]

[I 2025-04-24 16:10:41,286] Trial 21 finished with value: 0.7163244268331483 and parameters: {'n_estimators': 2544, 'max_depth': 5, 'learning_rate': 0.019554322843647925, 'subsample': 0.7336488866367249, 'colsample_bytree': 0.7812150129274054, 'reg_alpha': 3.088469185067023, 'reg_lambda': 5.7426479739572445}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  46%|████▌     | 23/50 [25:38<38:14, 84.99s/it]

[I 2025-04-24 16:12:44,348] Trial 22 finished with value: 0.711719850976119 and parameters: {'n_estimators': 2905, 'max_depth': 5, 'learning_rate': 0.03538892471967653, 'subsample': 0.7232843516761341, 'colsample_bytree': 0.8085549869107135, 'reg_alpha': 2.7562096787158783, 'reg_lambda': 6.526903111770548}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  48%|████▊     | 24/50 [27:41<41:48, 96.50s/it]

[I 2025-04-24 16:14:47,687] Trial 23 finished with value: 0.7136840408437302 and parameters: {'n_estimators': 2429, 'max_depth': 6, 'learning_rate': 0.012498826941922093, 'subsample': 0.7781765985838128, 'colsample_bytree': 0.9703098019351526, 'reg_alpha': 1.4345145529734316, 'reg_lambda': 5.291487941191829}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  50%|█████     | 25/50 [28:41<35:41, 85.66s/it]

[I 2025-04-24 16:15:48,079] Trial 24 finished with value: 0.7142303932266959 and parameters: {'n_estimators': 1729, 'max_depth': 4, 'learning_rate': 0.0618556631378607, 'subsample': 0.7362561272019728, 'colsample_bytree': 0.8551720843557867, 'reg_alpha': 2.428347651342529, 'reg_lambda': 4.106417757242812}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  52%|█████▏    | 26/50 [29:45<31:37, 79.07s/it]

[I 2025-04-24 16:16:51,769] Trial 25 finished with value: 0.7025544959500791 and parameters: {'n_estimators': 1912, 'max_depth': 5, 'learning_rate': 0.09808234909566142, 'subsample': 0.7011608246483444, 'colsample_bytree': 0.7463751722391186, 'reg_alpha': 3.8865247782102426, 'reg_lambda': 7.647219162152949}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  54%|█████▍    | 27/50 [30:17<24:52, 64.87s/it]

[I 2025-04-24 16:17:23,519] Trial 26 finished with value: 0.7175571572947226 and parameters: {'n_estimators': 951, 'max_depth': 6, 'learning_rate': 0.037917858134931655, 'subsample': 0.7334022059692845, 'colsample_bytree': 0.9691745676170529, 'reg_alpha': 0.914840767918407, 'reg_lambda': 5.288960725435194}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  56%|█████▌    | 28/50 [30:51<20:26, 55.76s/it]

[I 2025-04-24 16:17:58,007] Trial 27 finished with value: 0.7110342341653086 and parameters: {'n_estimators': 1512, 'max_depth': 4, 'learning_rate': 0.12622873734005205, 'subsample': 0.8682647813168562, 'colsample_bytree': 0.9165903231040327, 'reg_alpha': 5.139525663948572, 'reg_lambda': 6.455906454176148}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  58%|█████▊    | 29/50 [31:17<16:19, 46.63s/it]

[I 2025-04-24 16:18:23,352] Trial 28 finished with value: 0.7059236769082403 and parameters: {'n_estimators': 899, 'max_depth': 5, 'learning_rate': 0.07391661670428201, 'subsample': 0.7645066291284794, 'colsample_bytree': 0.921078215134362, 'reg_alpha': 2.28148136059073, 'reg_lambda': 4.578780455022366}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  60%|██████    | 30/50 [32:23<17:30, 52.52s/it]

[I 2025-04-24 16:19:29,596] Trial 29 finished with value: 0.7089303040774788 and parameters: {'n_estimators': 1223, 'max_depth': 8, 'learning_rate': 0.03294043983536622, 'subsample': 0.8072933993053406, 'colsample_bytree': 0.8744900279562509, 'reg_alpha': 0.01714045862249569, 'reg_lambda': 9.683803961885125}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  62%|██████▏   | 31/50 [32:53<14:32, 45.94s/it]

[I 2025-04-24 16:20:00,188] Trial 30 finished with value: 0.6951109859459974 and parameters: {'n_estimators': 622, 'max_depth': 7, 'learning_rate': 0.24716936556517627, 'subsample': 0.7428461845048825, 'colsample_bytree': 0.7577048694799419, 'reg_alpha': 1.2738790056236087, 'reg_lambda': 8.832237767858366}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  64%|██████▍   | 32/50 [33:21<12:06, 40.38s/it]

[I 2025-04-24 16:20:27,589] Trial 31 finished with value: 0.721487892202196 and parameters: {'n_estimators': 782, 'max_depth': 6, 'learning_rate': 0.011749060573894856, 'subsample': 0.7230057710807488, 'colsample_bytree': 0.8932196064490976, 'reg_alpha': 2.7104881005168338, 'reg_lambda': 6.113176571827363}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  66%|██████▌   | 33/50 [33:54<10:51, 38.33s/it]

[I 2025-04-24 16:21:01,136] Trial 32 finished with value: 0.7181283936188438 and parameters: {'n_estimators': 987, 'max_depth': 6, 'learning_rate': 0.026477188559521813, 'subsample': 0.7168494643824691, 'colsample_bytree': 0.8353003507527076, 'reg_alpha': 3.2043900220882557, 'reg_lambda': 5.5711143239658965}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  68%|██████▊   | 34/50 [34:15<08:48, 33.05s/it]

[I 2025-04-24 16:21:21,867] Trial 33 finished with value: 0.7220676689994291 and parameters: {'n_estimators': 725, 'max_depth': 5, 'learning_rate': 0.010964760339315854, 'subsample': 0.7014143706180264, 'colsample_bytree': 0.708444602449724, 'reg_alpha': 4.2866688465108975, 'reg_lambda': 6.8697657900830835}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  70%|███████   | 35/50 [34:33<07:06, 28.45s/it]

[I 2025-04-24 16:21:39,582] Trial 34 finished with value: 0.7227347061180381 and parameters: {'n_estimators': 621, 'max_depth': 5, 'learning_rate': 0.047510074929535726, 'subsample': 0.7742093822634245, 'colsample_bytree': 0.7218333909689377, 'reg_alpha': 4.510428927932546, 'reg_lambda': 7.48546863458625}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  72%|███████▏  | 36/50 [34:57<06:21, 27.23s/it]

[I 2025-04-24 16:22:03,973] Trial 35 finished with value: 0.7157730072614005 and parameters: {'n_estimators': 1316, 'max_depth': 3, 'learning_rate': 0.04897522018525305, 'subsample': 0.7777005461970019, 'colsample_bytree': 0.7331654145145192, 'reg_alpha': 7.000195814434967, 'reg_lambda': 7.325547069682667}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  74%|███████▍  | 37/50 [35:57<08:02, 37.14s/it]

[I 2025-04-24 16:23:04,225] Trial 36 finished with value: 0.7110852153181735 and parameters: {'n_estimators': 2733, 'max_depth': 4, 'learning_rate': 0.06233153467402474, 'subsample': 0.760718114782179, 'colsample_bytree': 0.7211348143394143, 'reg_alpha': 4.803531826644074, 'reg_lambda': 7.86344230412369}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  76%|███████▌  | 38/50 [36:54<08:34, 42.89s/it]

[I 2025-04-24 16:24:00,551] Trial 37 finished with value: 0.702563179160903 and parameters: {'n_estimators': 1874, 'max_depth': 5, 'learning_rate': 0.18096672482301818, 'subsample': 0.7851303218363057, 'colsample_bytree': 0.8031657627226304, 'reg_alpha': 3.8077716922016585, 'reg_lambda': 2.8766088588786025}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  78%|███████▊  | 39/50 [37:17<06:47, 37.06s/it]

[I 2025-04-24 16:24:23,994] Trial 38 finished with value: 0.6967475689745499 and parameters: {'n_estimators': 504, 'max_depth': 7, 'learning_rate': 0.09661266704358001, 'subsample': 0.7440381681625458, 'colsample_bytree': 0.7614741305074174, 'reg_alpha': 2.078876539998084, 'reg_lambda': 3.48133713469128}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  80%|████████  | 40/50 [37:57<06:17, 37.79s/it]

[I 2025-04-24 16:25:03,491] Trial 39 finished with value: 0.7142731193818734 and parameters: {'n_estimators': 2203, 'max_depth': 3, 'learning_rate': 0.0725324322423319, 'subsample': 0.8326787483329293, 'colsample_bytree': 0.7027671859813276, 'reg_alpha': 5.632383547902386, 'reg_lambda': 8.83177987041766}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  82%|████████▏ | 41/50 [38:19<04:59, 33.24s/it]

[I 2025-04-24 16:25:26,100] Trial 40 finished with value: 0.7162310050005303 and parameters: {'n_estimators': 1054, 'max_depth': 4, 'learning_rate': 0.04250706610313037, 'subsample': 0.8666614115123135, 'colsample_bytree': 0.8189267768866054, 'reg_alpha': 0.8471054655287253, 'reg_lambda': 6.18169989326697}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  84%|████████▍ | 42/50 [38:38<03:50, 28.84s/it]

[I 2025-04-24 16:25:44,670] Trial 41 finished with value: 0.7225762910471434 and parameters: {'n_estimators': 653, 'max_depth': 5, 'learning_rate': 0.026015221104305684, 'subsample': 0.7064093459655248, 'colsample_bytree': 0.7195134389004862, 'reg_alpha': 4.1249286166542785, 'reg_lambda': 6.7799809178035515}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  86%|████████▌ | 43/50 [38:56<02:58, 25.57s/it]

[I 2025-04-24 16:26:02,608] Trial 42 finished with value: 0.7255095272427056 and parameters: {'n_estimators': 624, 'max_depth': 5, 'learning_rate': 0.02540054371509027, 'subsample': 0.7252714166792115, 'colsample_bytree': 0.7343628536660826, 'reg_alpha': 4.435809063154365, 'reg_lambda': 6.917594786652922}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  88%|████████▊ | 44/50 [39:26<02:40, 26.83s/it]

[I 2025-04-24 16:26:32,370] Trial 43 finished with value: 0.7223323998762751 and parameters: {'n_estimators': 854, 'max_depth': 6, 'learning_rate': 0.03132437487099831, 'subsample': 0.7261040903535106, 'colsample_bytree': 0.7785826862264545, 'reg_alpha': 4.597659521678573, 'reg_lambda': 7.534698165058521}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  90%|█████████ | 45/50 [39:44<02:00, 24.20s/it]

[I 2025-04-24 16:26:50,437] Trial 44 finished with value: 0.7243214498954232 and parameters: {'n_estimators': 626, 'max_depth': 5, 'learning_rate': 0.054257284006339807, 'subsample': 0.7184198624452963, 'colsample_bytree': 0.740552899654459, 'reg_alpha': 3.486736880717915, 'reg_lambda': 8.421673484791425}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  92%|█████████▏| 46/50 [40:03<01:30, 22.64s/it]

[I 2025-04-24 16:27:09,428] Trial 45 finished with value: 0.7244229839880165 and parameters: {'n_estimators': 832, 'max_depth': 4, 'learning_rate': 0.05944045651877405, 'subsample': 0.7170325370555068, 'colsample_bytree': 0.7411304039170954, 'reg_alpha': 3.6297341823308433, 'reg_lambda': 8.381242400344723}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  94%|█████████▍| 47/50 [40:18<01:01, 20.35s/it]

[I 2025-04-24 16:27:24,458] Trial 46 finished with value: 0.7169158605435424 and parameters: {'n_estimators': 623, 'max_depth': 4, 'learning_rate': 0.1145681361391547, 'subsample': 0.71502792285806, 'colsample_bytree': 0.7369743971487093, 'reg_alpha': 3.5065464153779566, 'reg_lambda': 9.839972043841883}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  96%|█████████▌| 48/50 [40:34<00:38, 19.22s/it]

[I 2025-04-24 16:27:41,033] Trial 47 finished with value: 0.7189606258496759 and parameters: {'n_estimators': 849, 'max_depth': 3, 'learning_rate': 0.08288731562124517, 'subsample': 0.750959391959956, 'colsample_bytree': 0.7632600075153755, 'reg_alpha': 3.608543800387094, 'reg_lambda': 8.48348611406188}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775:  98%|█████████▊| 49/50 [41:00<00:21, 21.05s/it]

[I 2025-04-24 16:28:06,352] Trial 48 finished with value: 0.7146341268455808 and parameters: {'n_estimators': 1100, 'max_depth': 4, 'learning_rate': 0.061597749611161115, 'subsample': 0.7145414151320666, 'colsample_bytree': 0.9852409588197536, 'reg_alpha': 2.082755347179272, 'reg_lambda': 8.9863625202634}. Best is trial 11 with value: 0.7267751825365547.


Best trial: 11. Best value: 0.726775: 100%|██████████| 50/50 [41:31<00:00, 49.83s/it]

[I 2025-04-24 16:28:37,791] Trial 49 finished with value: 0.7100451569784157 and parameters: {'n_estimators': 828, 'max_depth': 6, 'learning_rate': 0.09272990190140357, 'subsample': 0.7363564816005026, 'colsample_bytree': 0.9528118969636823, 'reg_alpha': 3.1682063284417445, 'reg_lambda': 8.261279961517689}. Best is trial 11 with value: 0.7267751825365547.
,Лучший ROC-AUC: 0.7268
,Лучшие параметры: {'n_estimators': 653, 'max_depth': 6, 'learning_rate': 0.02176645480492359, 'subsample': 0.7021475108158397, 'colsample_bytree': 0.9160694536741326, 'reg_alpha': 2.5180991485275133, 'reg_lambda': 6.366469735868886}


In [10]:
best_params = study.best_params
best_params

NameError: name 'study' is not defined

Был сбой, но best_params удалось сохранить в save.txt так что скопирую их оттуда

In [13]:
best_params = {
    'n_estimators': 653,
    'max_depth': 6,
    'learning_rate': 0.02176645480492359,
    'subsample': 0.7021475108158397,
    'colsample_bytree': 0.9160694536741326,
    'reg_alpha': 2.5180991485275133,
    'reg_lambda': 6.366469735868886
}

In [14]:
# generate output data
X_true_test_df = pd.read_csv('./datasets/test/test.csv')
X_true_test = X_true_test_df.drop(columns=['index'])
indexes = X_true_test_df['index']

In [15]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [16]:
pipe = Pipeline([
    ('selector', VarianceThreshold(threshold=0.01)),
    ('scaler', StandardScaler()),
    ('model', XGBClassifier(**best_params, objective='binary:logistic', 
                          scale_pos_weight=10, device='cuda', tree_method='hist', eval_metric='auc')),
    
    # ('kbest', SelectKBest(f_classif, k=500)),
])

pipe.fit(X_train, y_train)

y_train_pred_proba = pipe.predict_proba(X_test)[:, 1]
roc_auc = roc_auc_score(y_test, y_train_pred_proba)
print(f'ROC-AUC is {roc_auc:.4f}')

c:\Users\ADMIN\AppData\Local\Programs\Python\Python312\Lib\site-packages\xgboost\core.py:729: UserWarning: [17:46:52] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\common\error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.
Potential solutions:
- Use a data structure that matches the device ordinal in the booster.
- Set the device for booster before call to inplace_predict.

This warning will only be shown once.

  return func(**kwargs)


ROC-AUC is 0.7253


In [17]:
y_test_pred_proba = pipe.predict_proba(X_true_test)[:, 1]
y_test_pred_proba

array([0.06112183, 0.09746365, 0.03243464, ..., 0.03522414, 0.13069998,
       0.02850325], shape=(76225,), dtype=float32)

In [20]:
output = pd.concat([indexes, pd.Series(y_test_pred_proba)], axis=1)
print(output)
output.to_csv("output.csv", index=False)

        index         0
0      124365  0.061122
1      177140  0.097464
2      106632  0.032435
3       17630  0.018898
4       97189  0.013972
...       ...       ...
76220   55074  0.046130
76221   12771  0.126945
76222   80999  0.035224
76223   31517  0.130700
76224  250238  0.028503

[76225 rows x 2 columns]


In [12]:
model = XGBClassifier(**best_params, objective='binary:logistic', 
                          scale_pos_weight=10, device='cuda', tree_method='hist', eval_metric='auc')


In [ ]:
X_true_test = pd.read_csv('./datasets/test/test.csv')
X_true_test_selected = pipe.fit

index  feature_0  feature_1  feature_2  feature_3  feature_4  \
0      124365   0.719921        1.0        1.0        1.0        1.0   
1      177140   0.123122        0.0        1.0        1.0        1.0   
2      106632   0.384612        0.0        1.0        1.0        1.0   
3       17630   0.452527        0.0        1.0        1.0        1.0   
4       97189   0.379611        1.0        1.0        1.0        2.0   
...       ...        ...        ...        ...        ...        ...   
76220   55074   0.325234        0.0        1.0        1.0        1.0   
76221   12771   0.227776        1.0        1.0        1.0        1.0   
76222   80999   0.459919        1.0        1.0        1.0        1.0   
76223   31517   0.667716        1.0        1.0        1.0        0.0   
76224  250238   0.226239        0.0        1.0        1.0        3.0   

       feature_5  feature_6  feature_7  feature_8  ...  feature_1357  \
0            1.0        2.0        1.0        0.0  ...           0.0   
1            2.0        1.0        2.0        0.0  ...           0.0   
2            0.0        1.0        2.0        0.0  ...           0.0   
3            4.0        5.0        2.0        0.0  ...           0.0   
4           11.0        3.0        4.0        0.0  ...           0.0   
...          ...        ...        ...        ...  ...           ...   
76220        3.0        1.0        0.0        3.0  ...           0.0   
76221        3.0        3.0        2.0        0.0  ...           0.0   
76222        2.0        1.0        2.0        3.0  ...           0.0   
76223        2.0        2.0        2.0        0.0  ...           0.0   
76224        2.0        1.0        0.0        4.0  ...           0.0   

       feature_1358  feature_1359  feature_1360  feature_1361  feature_1362  \
0               0.0           0.0             0             0             0   
1               0.0           0.0             0             0             0   
2               0.0           0.0             0             1             0   
3               0.0           0.0             0             0             0   
4               0.0           0.0             0             0             0   
...             ...           ...           ...           ...           ...   
76220           0.0           0.0             0             0             0   
76221           0.0           0.0             0             0             0   
76222           0.0           0.0             0             0             0   
76223           0.0           0.0             0             0             0   
76224           0.0           0.0             0             0             0   

       feature_1363  feature_1364  feature_1365  feature_1366  
0                 0             0             0             0  
1                 0             0             0             0  
2                 0             0             0             0  
3                 0             0             0             0  
4                 0             0             0             0  
...             ...           ...           ...           ...  
76220             0             0             0             0  
76221             0             0             0             0  
76222             0             0             0             0  
76223             0             0             0             0  
76224             0             0             0             0  

[76225 rows x 1368 columns]

In [15]:
model.fit(X_train, y_train)
y_train_pred = model.predict_proba(X_test)[:, 1]
roc_auc = roc_auc_score(y_test, y_train_pred)
print(f'ROC-AUC is {roc_auc:.4f}')

ROC-AUC is 0.7268


In [11]:
pipe = Pipeline([
    ('selector', VarianceThreshold(threshold=0.01)),
    # ('kbest', SelectKBest(f_classif, k=100)),
    ('scaler', StandardScaler()),
    ('xgb', XGBClassifier(  # XGBoost с настройками для ROC-AUC
        objective='binary:logistic',
        eval_metric='auc',
        n_estimators=2500,
        max_depth=6,
        learning_rate=0.01,
        subsample=0.8,
        colsample_bytree=0.4,
        reg_alpha=1,
        reg_lambda=1,
        random_state=42,
        tree_method='hist',
        device="cuda"
    ))
])

pipe.fit(X_train, y_train)
y_proba = pipe.predict_proba(X_test)[:, 1]
roc = roc_auc_score(y_test, y_proba)
roc

KeyboardInterrupt: 